From: Tim Douglas <me@timdoug.com>
Subject: [PATCH] nptl: convert time64 timer outputs to public itimerspec

The kernel uses 64-bit nanosecond fields while libc uses long. Passing
a public output buffer directly corrupts both its contents and adjacent
memory on 32-bit targets. Use kernel-layout temporaries and convert only
after successful syscalls, including the optional timer_settime output.

diff --git a/libpthread/nptl/sysdeps/unix/sysv/linux/timer_gettime.c b/libpthread/nptl/sysdeps/unix/sysv/linux/timer_gettime.c
index fccda69..c2778fa 100644
--- a/libpthread/nptl/sysdeps/unix/sysv/linux/timer_gettime.c
+++ b/libpthread/nptl/sysdeps/unix/sysv/linux/timer_gettime.c
@@ -23,6 +23,10 @@
 #include <bits/kernel-features.h>
 #include "kernel-posix-timers.h"
 
+#if defined(__UCLIBC_USE_TIME64__)
+#include "internal/time64_helpers.h"
+#endif
+
 
 #if defined(__NR_timer_gettime) || defined(__NR_timer_gettime64)
 # ifndef __ASSUME_POSIX_TIMERS
@@ -49,9 +53,18 @@ timer_gettime (
     {
       struct timer *kt = (struct timer *) timerid;
 
-      /* Delete the kernel timer object.  */
+      /* Query the kernel timer object.  */
 # if defined(__UCLIBC_USE_TIME64__) && defined(__NR_timer_gettime64)
-      int res = INLINE_SYSCALL (timer_gettime64, 2, kt->ktimerid, value);
+      struct __its64_struct kvalue;
+      int res = INLINE_SYSCALL (timer_gettime64, 2, kt->ktimerid,
+                                value ? &kvalue : NULL);
+      if (res == 0 && value)
+        {
+          value->it_interval.tv_sec = kvalue.interval_tv_sec;
+          value->it_interval.tv_nsec = kvalue.interval_tv_nsec;
+          value->it_value.tv_sec = kvalue.value_tv_sec;
+          value->it_value.tv_nsec = kvalue.value_tv_nsec;
+        }
 # else
       int res = INLINE_SYSCALL (timer_gettime, 2, kt->ktimerid, value);
 # endif
diff --git a/libpthread/nptl/sysdeps/unix/sysv/linux/timer_settime.c b/libpthread/nptl/sysdeps/unix/sysv/linux/timer_settime.c
index fca839d..b04b717 100644
--- a/libpthread/nptl/sysdeps/unix/sysv/linux/timer_settime.c
+++ b/libpthread/nptl/sysdeps/unix/sysv/linux/timer_settime.c
@@ -56,10 +56,18 @@ timer_settime (
     {
       struct timer *kt = (struct timer *) timerid;
 
-      /* Delete the kernel timer object.  */
+      /* Set the kernel timer object.  */
 # if defined(__UCLIBC_USE_TIME64__) && defined(__NR_timer_settime64)
+      struct __its64_struct kold;
       int res = INLINE_SYSCALL (timer_settime64, 4, kt->ktimerid, flags,
-				TO_ITS64_P(value), ovalue);
+                                TO_ITS64_P(value), ovalue ? &kold : NULL);
+      if (res == 0 && ovalue)
+        {
+          ovalue->it_interval.tv_sec = kold.interval_tv_sec;
+          ovalue->it_interval.tv_nsec = kold.interval_tv_nsec;
+          ovalue->it_value.tv_sec = kold.value_tv_sec;
+          ovalue->it_value.tv_nsec = kold.value_tv_nsec;
+        }
 # else
       int res = INLINE_SYSCALL (timer_settime, 4, kt->ktimerid, flags,
 				value, ovalue);
